compiler/optimizer/Inliner.cpp

```diff
@@ -5114,22 +5114,36 @@ bool TR_InlinerBase::inlineCallTarget2(TR_CallStack * callStack, TR_CallTarget *
             _disableTailRecursion = true;
          }
       }
-   else if (comp()->getOption(TR_EnableOSR) && tif->crossedBasicBlock() && !comp()->osrInfrastructureRemoved()
-       && (tif->resultNode() == NULL || tif->resultNode()->getReferenceCount() == 0))
-      {
-      /**
-       * In OSR, we need to split block even for cases without virtual guard. This is
-       * because in OSR a block with OSR point must have an exception edge to the osrCatchBlock
-       * of correct callerIndex. Split the block here so that the OSR points from callee
-       * and from caller are separated.
-       *
-       * This will not uncommon the return value if the block is split, instead it expects a temporary
-       * to have been generated so that nothing is commoned. This is valid under the current
-       * inliner behaviour, as the existance of a catch block will result in the required temporary being created.
-       */
-      TR::Block * blockOfCaller = previousBBStartInCaller->getNode()->getBlock();
-      TR::Block * blockOfCallerInCalleeCFG = nextBBEndInCaller->getNode()->getBlock()->split(callNodeTreeTop, callerCFG);
-      callerCFG->copyExceptionSuccessors(blockOfCaller, blockOfCallerInCalleeCFG);
+   else if (comp()->getOption(TR_EnableOSR) && !comp()->osrInfrastructureRemoved())
+      {
+      TR::Block * blockOfNextBBEndInCaller = nextBBEndInCaller->getNode()->getBlock();
+      TR_OSRMethodData *osrMethodData = comp()->getOSRCompilationData()->findOSRMethodData(comp()->getCurrentInlinedSiteIndex(), calleeSymbol);
+
+      if (osrMethodData && osrMethodData->getOSRCatchBlock())
+         {
+         /**
+          * In OSR, we need to split block even for cases without virtual guard. This is
+          * because in OSR a block with OSR point must have an exception edge to the osrCatchBlock
+          * of correct callerIndex. Split the block here so that the OSR points from callee
+          * and from caller are separated.
+          *
+          * This will not uncommon the return value if the block is split, instead it expects a temporary
+          * to have been generated so that nothing is commoned. This is valid under the current
+          * inliner behaviour, as the existance of a catch block will result in the required temporary being created.
+          */
+
+         debugTrace(tracer(), "Splitting block to have distint exception edges for callee %d\n", comp()->getCurrentInlinedSiteIndex());
+
+         TR::Block * blockOfCallerInCalleeCFG = blockOfNextBBEndInCaller->split(callNodeTreeTop, callerCFG);
+         callerCFG->copyExceptionSuccessors(blockContainingTheCall, blockOfCallerInCalleeCFG);
+         }
+      else
+         {
+         debugTrace(tracer(), "Copy OSR exception edge from caller block_%d to block_%d\n", blockContainingTheCall->getNumber(), blockOfNextBBEndInCaller->getNumber());
+
+         // Need to copy the exception edge from caller block containing the call to the last callee block after merge
```

Comment:
could you add some comments to explain that `blockOfNextBBEndInCaller ` is merged with the last block in inlined callee so it loses all the original exception edges and therefore we need to copy those exception edges back to this block. To people not familiar with the code, `last callee block after merge` seems not clear enough to justify why we want to copy the exception edges from caller to callee.

Reply:
Made some changes to improve the comment and code readability.